### ISSUE TITLE
release 2.6.9 strange failure tring to do simple math function

### ISSUE THREAD

**Issue by tringate** [Program-O/Program-O#370]
I have discovered a strange failure using the following AIML test case category (FAILURE) trying to figure out why my login AIML code stopped working..

```
<category><pattern>FAILURE</pattern><template>
<condition name="logincount">
<li value="undefined"><set name="logincount">1</set></li>
<li><set name="logincount"><math><get name="logincount"/> add 1</math></set></li>
</condition>
</template></category>

<category><pattern>FAILURE1</pattern><template>
<math>1 add 1</math>
</template></category>`

There are actually two failures with this.  The first is that the condition should find the non-existent condition LOGINCOUNT and set it to a value of 1.  However even though that variable does not exist in the client_properties, it apparently does not execute that line of the condition but rather tries to execute the second line of the condition which tries to add 1 to whatever the value is and it throws an "Invalid math function!" error.  This might be a valid response if the value is not assigned at all.  In this particular case an unassigned value might be assumed to be a ZERO and let the math add 1 to 0.  I then tried the even simpler atomic math AIML test case (FAILURE1) which failed to add 1 to 1 with the same error.

So both CONDITION and the MATH tags are broken in release 2.6.9.

I know the MATH tag is not part of AIML 1.1, but it is distributed with Program-O and should be fixed.

To clarify the condition failure, this is using the fixed version of the condition tag Dave and found a few nights ago by changing line 823 in file "parse_aiml_as_XMP.php" to say:
            $client_property = (isset($client_properties[$name])) ? $client_properties[$name] : 'undefined';
I believe it was previously returning "notset" or something like that.

These tests are being run using a virgin copy of Program-O and not my modified version for the "genie" project.  I did try using both the upper and lower case of ADD.  That should not make a difference, and it didn't, they both failed.

**Comment by tringate** [comment]
With the push of the dev code dated 3/1/2018 the math part of this bug appears to be fixed.  The condition failure looks like the following to me:
If you use condition to test a variable and it is UNDEFINED then within that condition statement you are not able to set that same variable to be defined as a value.  This is a guess at the bug because that is what appears to be happening.
```
<category><pattern>FAILURE</pattern><template>
<condition name="logincount">
<li value="something">*** FAILED ***</li>
<li value="undefined">logincount was not defined, setting it to: <set name="logincount">1</set></li>
<li><set name="logincount"><math><get name="logincount"/> ADD 1</math></set></li>
</condition>
</template></category>`

The above test category always resolves as the UNDEFINED case, and the variable of LOGINCOUNT is never created in the database.

You can run this AIML category over and over again and the variable is never created hence it is never incremented to count the number of logins.

In my original posting of this bug I had a category of FAILURE1 which now works if you use UPPERCASE for the math operations of ADD SUBTRACT MULTIPLE and DIVIDE.  My preference would be that the operation name be case insensitive.

**Comment by tringate** [comment]
Bug fixed in dev 2.6.10
